test/framework/src/main/java/org/elasticsearch/test/NodeRoles.java

```diff
@@ -0,0 +1,187 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.test;
+
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.node.NodeRoleSettings;
+
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Utility methods for creating {@link Settings} instances defining a set of {@link DiscoveryNodeRole}.
+ */
+public class NodeRoles {
+
+    public static Settings onlyRole(final DiscoveryNodeRole role) {
+        return onlyRole(Settings.EMPTY, role);
+    }
+
+    public static Settings onlyRole(final Settings settings, final DiscoveryNodeRole role) {
+        return onlyRoles(settings, Set.of(role));
+    }
+
+    public static Settings onlyRoles(final Set<DiscoveryNodeRole> roles) {
+        return onlyRoles(Settings.EMPTY, roles);
+    }
+
+    public static Settings onlyRoles(final Settings settings, final Set<DiscoveryNodeRole> roles) {
+        return Settings.builder()
+            .put(settings)
+            .putList(
+                NodeRoleSettings.NODE_ROLES_SETTING.getKey(),
+                roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.toUnmodifiableList()))
+            .build();
+    }
+
+    public static Settings removeRoles(final Set<DiscoveryNodeRole> roles) {
+        return removeRoles(Settings.EMPTY, roles);
+    }
+
+    public static Settings removeRoles(final Settings settings, final Set<DiscoveryNodeRole> roles) {
+        final Settings.Builder builder = Settings.builder().put(settings);
+        builder.putList(
+            NodeRoleSettings.NODE_ROLES_SETTING.getKey(),
+            NodeRoleSettings.NODE_ROLES_SETTING.get(settings)
+                .stream()
+                .filter(Predicate.not(roles::contains))
+                .map(DiscoveryNodeRole::roleName)
+                .collect(Collectors.toUnmodifiableList())
+        );
+        return builder.build();
+    }
+
+    public static Settings addRoles(final Set<DiscoveryNodeRole> roles) {
+        return addRoles(Settings.EMPTY, roles);
+    }
+
+    public static Settings addRoles(final Settings settings, final Set<DiscoveryNodeRole> roles) {
+        final Settings.Builder builder = Settings.builder().put(settings);
+        builder.putList(
+            NodeRoleSettings.NODE_ROLES_SETTING.getKey(),
+            Stream.concat(NodeRoleSettings.NODE_ROLES_SETTING.get(settings).stream(), roles.stream())
+                .map(DiscoveryNodeRole::roleName)
+                .distinct()
```

Comment:
I can't help wondering if the roles setting implementation should be using a `Set` to begin with. 🤷‍♂️